File: auth/src/clients.ts
import type { Auth } from "./auth.js";
import { log } from "./log.js";

/**
 * The OAuth clients the operator registered, as the operator manages them
 * afterwards.
 *
 * Registering a client is standard and needs nothing from this file: RFC 7591
 * is what the OAuth provider plugin serves at `/oauth2/register`, and it is
 * the whole of the contract docs/AUTH.md asks an issuer for. Changing one
 * afterwards is RFC 7592, a separate specification the plugin does not
 * implement — its registration answer names no client configuration endpoint
 * and hands out no registration access token, so there is nothing standard
 * for the operator to call.
 *
 * That matters because the redirect list of an application's client is not
 * written once. It follows the project's environments: a preview appears with
 * a URL nobody could have written down in advance, and a merged pull request
 * takes that URL away again. Keeping the list in step is the point of
 * `ResourceClaim` type `oidcClient` (issue #8), and the operator is the only
 * component that knows when a URL comes or goes.
 *
 * So this rides where the account directory rides — the operator's own prefix
 * on the identity provider the chart ships, authenticated by its service
 * credential — for the same reason the directory does: the platform needs
 * something OpenID Connect never offered. An installation federated to an
 * issuer of its own keeps registration and loses the maintenance, which the
 * operator reports on the claim rather than papering over.
 *
 * **The operator may only manage what the operator registered.** Every client
 * the dynamic-registration endpoint creates records the account whose session
 * created it, which for the operator is the service account. The seeded
 * Kitchen UI client records nobody, because it is written straight into the
 * table at start-up. Requiring the owner to match therefore refuses a call
 * that would rewrite the dashboard's own redirect URIs or take its client
 * away — the one mistake in this file that would sign every person out of the
 * platform and leave nobody able to sign back in.
 */

/** One client, as the operator reads it back. Never its secret. */
export interface ManagedClient {
	clientId: string;
	name: string;
	redirectURIs: string[];
}

/** What a client may have changed about it. */
export interface ClientUpdate {
	clientName?: string;
	redirectURIs?: string[];
	grantTypes?: string[];
	scopes?: string[];
}

/** Thrown when the client exists but was not registered by the operator. */
export class NotOurClientError extends Error {}

/** The columns this module reads off an `oauthClient` row. */
interface ClientRow {
	id: string;
	clientId: string;
	name?: string | null;
	userId?: string | null;
	redirectUris?: string[] | string | null;
}

/**
 * A `string[]` column, whichever way the adapter hands it back. The Postgres
 * adapter answers with an array; some store it joined, and a client whose
 * redirect list read back as the characters of a string would be a client
 * nobody could sign in to.
 */
function asList(value: string[] | string | null | undefined): string[] {
	if (Array.isArray(value)) {
		return value;
	}
	return String(value ?? "")
		.split(",")
		.filter(Boolean);
}

function asManagedClient(row: ClientRow): ManagedClient {
	return {
		clientId: row.clientId,
		name: row.name ?? "",
		redirectURIs: asList(row.redirectUris),
	};
}

/** The client with this id, or null when there is none. */
async function findClient(auth: Auth, clientId: string): Promise<ClientRow | null> {
	const ctx = await auth.$context;
	return ctx.adapter.findOne<ClientRow>({
		model: "oauthClient",
		where: [{ field: "clientId", value: clientId }],
	});
}

/**
 * The client, established to be one this operator registered.
 *
 * Null means there is no such client — which the caller answers with a 404,
 * the same answer an issuer that had never heard of this endpoint would give,
 * and the same one the operator reads as "it is not there".
 */
async function ourClient(auth: Auth, operatorId: string, clientId: string): Promise<ClientRow | null> {
	const client = await findClient(auth, clientId);
	if (!client) {
		return null;
	}
	if (!client.userId || client.userId !== operatorId) {
		throw new NotOurClientError(
			`the client ${clientId} was not registered by the operator: it is managed where it was created`,
		);
	}
	return client;
}

/**
 * Replaces what the operator maintains about one of its clients.
 *
 * Only the named fields move. The client id, the secret and everything the
 * registration decided stay as they are: this is the redirect list following
 * the project's environments, not a re-registration, and a re-registration
 * would hand out credentials the running application does not have.
 */
export async function updateClient(
	auth: Auth,
	operatorId: string,
	clientId: string,
	update: ClientUpdate,
): Promise<ManagedClient | null> {
	const client = await ourClient(auth, operatorId, clientId);
	if (!client) {
		return null;
	}

	const changes: Record<string, unknown> = { updatedAt: new Date() };
	if (update.redirectURIs) {
		changes.redirectUris = update.redirectURIs;
	}
	if (update.clientName) {
		changes.name = update.clientName;
	}
	if (update.grantTypes?.length) {
		changes.grantTypes = update.grantTypes;
	}
	if (update.scopes?.length) {
		changes.scopes = update.scopes;
	}

	const ctx = await auth.$context;
	const updated = await ctx.adapter.update<ClientRow>({
		model: "oauthClient",
		where: [{ field: "id", value: client.id }],
		update: changes,
	});
	log.info("updated an operator-registered OAuth client", {
		clientId,
		redirectURIs: update.redirectURIs?.length ?? asList(client.redirectUris).length,
	});
	return asManagedClient(updated ?? { ...client, ...(update.redirectURIs ? { redirectUris: update.redirectURIs } : {}) });
}

/**
 * Deregisters one of the operator's clients.
 *
 * The consents and tokens that reference it are the plugin's rows, and they
 * are removed with it: a client id nothing can be issued for is exactly what
 * "the claim is gone" has to mean, since anything still holding a token would
 * otherwise keep signing people in with credentials the platform no longer
 * knows about.
 */
export async function deleteClient(
	auth: Auth,
	operatorId: string,
	clientId: string,
): Promise<ManagedClient | null> {
	const client = await ourClient(auth, operatorId, clientId);
	if (!client) {
		return null;
	}

	const ctx = await auth.$context;
	for (const model of ["oauthAccessToken", "oauthRefreshToken", "oauthConsent"]) {
		try {
			await ctx.adapter.deleteMany({
				model,
				where: [{ field: "clientId", value: clientId }],
			});
		} catch (error) {
			// A model this plugin version does not have is not a reason to
			// leave the client registered; the client row is what makes the
			// credentials usable, and it goes below either way.
			log.debug("nothing to remove for the client", { model, clientId, error: String(error) });
		}
	}
	await ctx.adapter.delete({ model: "oauthClient", where: [{ field: "id", value: client.id }] });
	log.info("deregistered an operator-registered OAuth client", { clientId });
	return asManagedClient(client);
}

File: internal/idp/clients.go
/*
Copyright 2026.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package idp

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

// ClientsPath is the operator's client-management endpoint on the identity
// provider the chart ships: the redirect list of a client it registered, and
// taking that client away again.
//
// Registering a client is standard — RFC 7591, and the whole of what
// docs/AUTH.md asks an issuer for. *Changing* one is RFC 7592, which is a
// separate specification an issuer may implement or not, and the one the
// chart ships does not: its registration response names no client
// configuration endpoint. So this sits where the account directory sits, for
// the same reason — under a prefix that is Kitchen's, authenticated by the
// operator's service credential — and it is reached only when the issuer
// offered nothing standard. An installation federated to an issuer of its own
// keeps every client it registers; what it does not keep is the redirect list
// being maintained for it. See ErrNoClientManagement.
const ClientsPath = "/kitchen/clients"

// ErrNoClientManagement says the issuer offers no way to change or remove a
// client once it has been registered: no RFC 7592 client configuration
// endpoint in its registration answer, and no Kitchen prefix either.
//
// It is a federated issuer, and it is not a fault to be fixed. Everything
// else works — the client is registered, the application signs people in with
// it — and what is lost is the automation on top:
//
//   - **A redirect list stops following the environments.** A preview that
//     appears after the client was registered has a callback URL the issuer
//     will refuse, so signing in from that preview fails until somebody adds
//     it at the issuer by hand. The claim says so on its condition rather
//     than looking bound and behaving otherwise.
//   - **A deleted claim leaves its client behind.** The operator says which
//     client id it could not remove, because a client nobody deregistered is
//     a credential that outlives the thing it was for.
//
// Callers report it rather than retry it: an endpoint that is not there does
// not appear by being asked again.
var ErrNoClientManagement = errors.New("the issuer offers no way to manage a registered client")

// ClientHandle is a registered client and how to get back to it.
//
// The two registration fields are RFC 7592's: an issuer that supports client
// management hands out a URL for the client and a token that authorizes
// changes to it, once, in the registration answer. They are stored with the
// client's credentials because they cannot be asked for again.
type ClientHandle struct {
	// ID is the client_id.
	ID string

	// RegistrationURI is the client configuration endpoint the issuer named,
	// empty when it named none.
	RegistrationURI string

	// RegistrationToken authorizes calls to RegistrationURI.
	RegistrationToken string
}

// Manageable reports whether the issuer named a client configuration
// endpoint for this client.
func (h ClientHandle) Manageable() bool {
	return h.RegistrationURI != "" && h.RegistrationToken != ""
}

// clientRequest is one client's metadata as both RFC 7591 registration and
// RFC 7592 update carry it.
func clientRequest(want ClientRegistration) map[string]any {
	payload := map[string]any{
		"client_name":                want.Name,
		"redirect_uris":              want.RedirectURIs,
		"grant_types":                want.GrantTypes,
		"response_types":             []string{"code"},
		"token_endpoint_auth_method": "client_secret_basic",
	}
	if len(want.Scopes) > 0 {
		payload["scope"] = strings.Join(want.Scopes, " ")
	}
	return payload
}

// UpdateClient replaces a registered client's metadata — in practice its
// redirect list, which is the one part of it the platform keeps moving.
//
// It prefers the standard route: RFC 7592 replaces the whole registration, so
// the caller passes what it would have registered rather than a patch. When
// the issuer named no client configuration endpoint it falls back to the
// Kitchen prefix, and when that is not there either the answer is
// ErrNoClientManagement — the caller's cue to report rather than retry.
func (c *Client) UpdateClient(ctx context.Context, handle ClientHandle, want ClientRegistration) error {
	if handle.ID == "" {
		return errors.New("updating an OAuth client: no client id")
	}
	what := fmt.Sprintf("updating the OAuth client %q", handle.ID)

	if handle.Manageable() {
		payload := clientRequest(want)
		payload["client_id"] = handle.ID
		body, err := json.Marshal(payload)
		if err != nil {
			return err
		}
		return c.configureClient(ctx, http.MethodPut, handle, body, what)
	}

	body, err := json.Marshal(map[string]any{
		"clientId":     handle.ID,
		"clientName":   want.Name,
		"redirectURIs": want.RedirectURIs,
		"grantTypes":   want.GrantTypes,
		"scopes":       want.Scopes,
	})
	if err != nil {
		return err
	}
	_, err = c.callDirectory(ctx, http.MethodPut, c.cfg.BaseURL+ClientsPath, body, what)
	if errors.Is(err, errDirectoryNotFound) {
		return fmt.Errorf("%s: %w", what, ErrNoClientManagement)
	}
	return err
}

// DeleteClient deregisters a client, so that credentials which have outlived
// what they were issued for cannot be used to sign anybody in.
//
// A client that is already gone is a success: the caller is a finalizer, and
// deletion asking for something twice has to be as good as asking once.
func (c *Client) DeleteClient(ctx context.Context, handle ClientHandle) error {
	if handle.ID == "" {
		return nil
	}
	what := fmt.Sprintf("deregistering the OAuth client %q", handle.ID)

	if handle.Manageable() {
		err := c.configureClient(ctx, http.MethodDelete, handle, nil, what)
		if errors.Is(err, errClientGone) {
			return nil
		}
		return err
	}

	endpoint := c.cfg.BaseURL + ClientsPath + "?" + url.Values{"clientId": []string{handle.ID}}.Encode()
	_, err := c.callDirectory(ctx, http.MethodDelete, endpoint, nil, what)
	if errors.Is(err, errDirectoryNotFound) {
		// The prefix answers 404 both for a client that is not there and for
		// an issuer that has never heard of it, and the two are the same
		// answer to a caller that is trying to remove the thing: it is not
		// there, and this is not the place to find out why.
		return nil
	}
	return err
}

// configureClient performs one RFC 7592 call against the client's own
// configuration endpoint, which is authorized by the registration access
// token rather than by the operator's service credential — the token is the
// issuer's statement about this one client.
func (c *Client) configureClient(
	ctx context.Context,
	method string,
	handle ClientHandle,
	body []byte,
	what string,
) error {
	req, err := c.request(ctx, method, handle.RegistrationURI, bytesReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("accept", "application/json")
	if body != nil {
		req.Header.Set("content-type", "application/json")
	}
	req.Header.Set("authorization", "Bearer "+handle.RegistrationToken)

	res, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("%s: %w", what, err)
	}
	defer func() { _ = res.Body.Close() }()

	answer, err := readLimited(res.Body)
	if err != nil {
		return err
	}
	switch {
	case res.StatusCode == http.StatusNotFound:
		return fmt.Errorf("%s: %w", what, errClientGone)
	case res.StatusCode < 200 || res.StatusCode > 299:
		return fmt.Errorf("%s: %s: %s", what, res.Status, summarize(answer))
	}
	// The answer is the client as the issuer now holds it, and nothing here
	// reads it: the operator's record is what it registered, and a client the
	// issuer described differently would be a client to register again rather
	// than one to believe.
	return nil
}

// errClientGone is the unwrapped 404 from a client configuration endpoint:
// the client this handle names is not registered any more.
var errClientGone = errors.New("the client is not registered")
